File: Java/Move Pieces to Obtain a String.java
public class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public boolean canChange(String start, String target) {
        int n = start.length();
        int i = 0, j = 0;

        while (i < n || j < n) {
            while (i < n && start.charAt(i) == SKIP) ++i;
            while (j < n && target.charAt(j) == SKIP) ++j;

            if (i == n && j == n)
                return true;

            if (i == n || j == n)
                return false;

            if (start.charAt(i) != target.charAt(j)) return false;
            if (start.charAt(i) == LEFT && i < j) return false;
            if (start.charAt(i) == RIGHT && i > j) return false;

            ++i;
            ++j;
        }

        return true;
    }

    public static final char SKIP = '_';
    public static final char LEFT = 'L';
    public static final char RIGHT = 'R';
}

File: Python/Move Pieces to Obtain a String.py
class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    SKIP = '_'
    LEFT = 'L'
    RIGHT = 'R'

    def canChange(self, start: str, target: str) -> bool:
        n = len(start)
        i = j = 0

        while i < n or j < n:
            while i < n and start[i] == self.SKIP:
                i += 1

            while j < n and target[j] == self.SKIP:
                j += 1

            if i != j and (i == n or j == n):
                return False

            if i == j == n:
                return True

            if i == n or j == n:
                return False

            if start[i] != target[j]:
                return False

            if start[i] == self.LEFT and i < j:
                return False

            if start[i] == self.RIGHT and i > j:
                return False

            i += 1
            j += 1

        return True
